### PR TITLE
Node compatibility

### PR DESCRIPTION
**Pull request by kaicataldo**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -4,7 +4,6 @@ import Engine from './src/Engine';
 
 var Focss = Class.extend({
   init: function(root, extensions, scoped) {
-    scoped = scoped === undefined ? true : !!scoped;
```
Comment (kaicataldo) [inline index.js]: This check is now done in the Style constructor function

File: src/Engine.js [hunk 2]
```diff
@@ -280,14 +261,14 @@ Engine = Class.extend({
...
+      this.style.insertRule(':root {}', i);
     }
     else {
-      this.sheet.insertRule(rule.getSelector(this._prefix) + '{}', i);
+      this.style.insertRule(rule.getSelector(this.style.prefix) + '{}', i);
     }
     this.rules.push(rule);
 
-    if (this._state) {
+    if (this.style.hasStyleSheet && this._state) {
```
Comment (kaicataldo) [inline src/Engine.js]: We may want to consider *not* auto processing on insert in the future, but this guard allows us to not a make this a breaking change.

File: test/behavior/toString.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import Focss from '../..';
```
Comment (kaicataldo) [inline test/behavior/toString.js]: We had discussed not taking time to write tests for a Node environment in this PR, but I found that the quickest way to figure out what was failing in Node was just to run it. Since I had the tests anyway, figured I should include them. These three tests hit the three forks of logic we currently have for inserting Focss descriptors.

File: util/css.js [hunk 4]
```diff
@@ -20,27 +20,11 @@ var isUnitlessNumber = {
...
-    }
-  }
-}(Element.prototype));
-
 export default {
   find: function(selector, base) {
     return (base || document).querySelectorAll(selector);
   },
 
-  matches: function(element, selector) {
```
Comment (kaicataldo) [inline util/css.js]: This isn't used anywhere

File: util/css.js [hunk 5]
```diff
@@ -96,6 +80,7 @@ export default {
       result += key + ':' + value + 'px;';
     }
     result += '}';
-    return result;
+
+    return result.trim();
```
Comment (kaicataldo) [inline util/css.js]: Hadn't tested for multiple rules and noticed there are spaces between them. This only trims the whitespace around the entire rule (selector + body)

File: util/detectScoped.js [hunk 6]
```diff
@@ -1,59 +1,61 @@
...
-  // we need to append it to the DOM because the DOM element at least FF keeps NULL as a sheet utill appended
-  // and we can't check for the rules / cssRules and changeSelectorText untill we have that
-  document.body.appendChild(check);
-  testSheet = check[DOMStyle];
-
-  // add a test styleRule to be able to test selectorText changing support
-  // IE doesn't allow inserting of '' as a styleRule
-  if (testSheet.addRule) {
-    testSheet.addRule('c', 'blink');
+export default {
```
Comment (kaicataldo) [inline util/detectScoped.js]: Changed from an IIFE so that we can conditionally detect scoped based on if the styles are being created in the browser or in a Node env

File: src/Style.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
+import detectScoped from '../util/detectScoped';
+
+function genId() {
+  return '__focss__' + genId.i++;
+}
+genId.i = 1;
+
+export default class Style {
```
Comment (mikesherov) [inline src/Style.js]: Null object pattern? Not in love.

File: test/behavior/toString.js (hunk 3 above)
Comment (mikesherov) [inline test/behavior/toString.js]: ok, cool :-)

File: test/unit/index.js [hunk 8]
```diff
@@ -50,6 +50,8 @@ describe('Focss', function() {
 
   describe('#toString()', function() {
     it('returns a string of processed styles', function() {
+      fox = new Focss();
```
Comment (mikesherov) [inline test/unit/index.js]: in the main afterEach, put `fox = null;`, it would've vcaught this bug

File: src/Engine.js [hunk 9]
```diff
@@ -280,14 +261,14 @@ Engine = Class.extend({
 
     if (rule.isComputed) {
       // Placeholder rule
-      this.sheet.insertRule(':root {}', i);
+      this.style.insertRule(':root {}', i);
     }
     else {
-      this.sheet.insertRule(rule.getSelector(this._prefix) + '{}', i);
+      this.style.insertRule(rule.getSelector(this.style.prefix) + '{}', i);
```
Comment (mikesherov) [inline src/Engine.js]: I though we were nuking concept of prefix and scoped?

File: src/Engine.js [hunk 10]
```diff
@@ -127,17 +106,19 @@ Engine = Class.extend({
...
-
-    // Selector has changed
-    if (selector !== this.cssRules[i].selectorText) {
-      this.sheet.deleteRule(i);
-      this.sheet.insertRule(rule.getSelector(this._prefix) + '{}', i);
-      result = true;
-    }
+    var selector = rule.getSelector(this.style.prefix);
+
+    if (this.style.hasStyleSheet) {
```
Comment (mikesherov) [inline src/Engine.js]: this is a weak inference. Better to say `isBrowser`

Comment (mikesherov) [inline src/Engine.js]: or even, `hasDom`

File: src/Style.js [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    if (!scoped || this.compat.scopeSupported || root === document.body) {
+      this._prefix = '';
+    }
+    else {
+      this._prefix = '#' + (root.id || (root.id = genId())) + ' ';
+    }
+  }
+
+  insertRule(rule, index) {
+    if (!this.sheet) { return; }
```
Comment (mikesherov) [inline src/Style.js]: why is this needed?

File: src/Style.js [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+    return this.sheet.insertRule(rule, index);
+  }
+
+  deleteRule(index) {
+    if (!this.sheet) { return; }
```
Comment (mikesherov) [inline src/Style.js]: why is this needed?

File: src/Style.js [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+    if (!this.style) { return; }
+    return this.style[this.compat.sheet];
+  }
+
+  get cssRules() {
+    if (!this.sheet) { return; }
+    return this.sheet[this.compat.rules];
+  }
+
+  get rulesLength() {
```
Comment (mikesherov) [inline src/Style.js]: not a fan of all these getters.

File: src/Engine.js (hunk 10 above)
Comment (mikesherov) [inline src/Engine.js]: and shouldn't this be a gaurd clause at the top instead?

Comment (mikesherov) [inline src/Engine.js]: like it seems cleaner to put `typeof navigator` check at the top of this function.

Comment (mikesherov) [inline src/Engine.js]: actually... at the top of `process`

File: src/Engine.js [hunk 14]
```diff
@@ -29,40 +24,24 @@ var foreachSelectorRegex = /%forEach\(([^,]+),(.+)\)$/i;
 var filterEachSelectorRegex = /%filterEach\(([^,]+),([^,]+),(.+)\)$/i;
 var toggleSelectorPsuedoRegex = /:(hover|active)/;
 var toggleSelectorClassRegex = /\.(__[^ :]+)/;
-var otherPrefixRegex = getVendorPrefixRegex();
+var otherPrefixRegex = typeof navigator === 'undefined' ? new RegExp() : getVendorPrefixRegex();
```
Comment (hzoo) [inline src/Engine.js]: Do we want a comment here to explain that the check is for node compat?

File: src/Engine.js (hunk 2 above)
Comment (mikesherov) [inline src/Engine.js]: need to auto process on insert
although I guess that's not a real use case?

File: src/Engine.js (hunk 9 above)
Comment (kaicataldo) [inline src/Engine.js]: Oh, okay - happy to do that, but wasn't aware. Sounds good to me :+1:

File: src/Style.js (hunk 11 above)
Comment (kaicataldo) [inline src/Style.js]: The idea for these was for Engine to not have to care about whether the style sheet existed or not - just returns a No op in the case that there is no DOM

File: src/Engine.js (hunk 10 above)
Comment (kaicataldo) [inline src/Engine.js]: Ah, you're right - I was trying to guard against `_process` being called by `insert` when Focss doesn't have access to the DOM but there's another guard in place for that so this is totally unnecessary. Makes sense to do the check in `process` anyway, though.

**Comment by kaicataldo** [comment]
Okay, so rather than the null object pattern, I could also add a check into Engine like: 

```
function hasDom() {
  return typeof window !== 'undefined';
}
```

Then I could use that for all the guards/working around DOM-reliant behavior, including bailing on `process` immediately if Focss doesn't have access to the DOM.

Two questions:
 * Thoughts on keeping the Style class? Or do you think that should all be inlined into the Engine? I like the separation (easier for me to reason about).
 * The other idea @Attamusc and I discussed was not auto-processing on insert when there is state, but that's a breaking change and I felt like that should be a separate issue if we decide to do it.

File: src/Engine.js [hunk 15]
```diff
@@ -147,7 +128,7 @@ Engine = Class.extend({
     // ignore rules that contain the other vendor prefix, as trying to
     // insert them into a stylesheet will cause an exception to be thrown
     // @see: http://stackoverflow.com/questions/23050001/insert-multiple-css-rules-into-a-stylesheet
-    if (otherPrefixRegex.test(selector)) {
+    if (this.style.hasStyleSheet && otherPrefixRegex.test(selector)) {
```
Comment (mikesherov) [inline src/Engine.js]: this file engine should know if it's in a browser or not. it shouldn't just infer from stylesheet

Comment (mikesherov) [inline src/Engine.js]: or at least defer to `this.style.getPrefixRegex`

File: test/unit/index.js [hunk 16]
```diff
@@ -1,24 +1,39 @@
...
+      this._fox = null;
+    }
+  });
 
   it('is instanciable', function() {
-    expect(function() {
+    let fox;
+
+    expect(() => {
       fox = new Focss();
```
Comment (kaicataldo) [inline test/unit/index.js]: This instance needs to be managed independently so that we ensure that the stylesheet that is created is also cleaned up from the DOM - the current implementation of this test leaks state into later tests

File: test/unit/src/Engine.js [hunk 17]
```diff
@@ -13,9 +13,14 @@ describe('Engine', function() {
...
   it('is a constructor', function() {
+    let engine;
+
     expect(function() {
-      new Engine();
+      engine = new Engine();
     }).not.toThrow();
+    expect(engine).toBeDefined();
+
+    engine.destroy();
```
Comment (kaicataldo) [inline test/unit/src/Engine.js]: Same as https://github.com/behance/focss/pull/37/files#r72528747

File: src/Engine.js [hunk 18]
```diff
@@ -299,33 +292,6 @@ Engine = Class.extend({
...
-.mixin({
-  get sheet() {
-    return this.style[compat.sheet];
-  },
-  get cssRules() {
-    return this.sheet[compat.rules];
-  }
 });
 
-// ES6 future-proofing
```
Comment (kaicataldo) [inline src/Engine.js]: We can get rid of this now that we have ES6 with Babel, right?

File: src/Engine.js (hunk 14 above)
Comment (kaicataldo) [inline src/Engine.js]: Using `hasDom` now, so I think it's self-documenting

**Comment by kaicataldo** [comment]
Updated - Style is a simple wrapper around the stylesheet now and Engine handles all the DOM-checking/management of the stylesheet. Also removed scoped styles/prefixes, as per comments.

File: src/Engine.js (hunk 18 above)
Comment (mikesherov) [inline src/Engine.js]: No. We need this code so that Engine is iterable. What *could* be removed is just the `typeof Symbol` check. Not worth removing IMO

File: src/Style.js [hunk 19]
```diff
@@ -0,0 +1,31 @@
+import detectScoped from '../util/detectScoped';
+
+export default class Style {
+  constructor(root = document.body) {
```
Comment (mikesherov) [inline src/Style.js]: Why is there a default if root is always being passed in?

File: src/Rule.js [hunk 20]
```diff
@@ -37,12 +36,8 @@ var Rule = Class.extend({
     }
   },
 
-  getSelector: function(prefix) {
-    if (!prefix) { return this.computedSelector || this.selector; }
-
-    return this.specificity.map(function(part) {
```
Comment (mikesherov) [inline src/Rule.js]: Do we still need specificity?

File: src/Engine.js (hunk 18 above)
Comment (kaicataldo) [inline src/Engine.js]: Got it

File: src/Rule.js (hunk 20 above)
Comment (kaicataldo) [inline src/Rule.js]: Isn't this only necessary when there's the concept of a prefix (which I removed in this PR)?

File: src/Style.js (hunk 19 above)
Comment (kaicataldo) [inline src/Style.js]: This matches the current behavior - I removed it initially because I was using whether or not a `root` parameter was passed in to determine if it was using the DOM or not. Now that we have the much more resilient `hasDom` check happening in `Engine`, we can have a default. That being said, I don't think we actually use it this way anywhere. Think I should get rid of it? Breaking chance, so figured I'd leave it in for now.

File: src/Rule.js (hunk 20 above)
Comment (mikesherov) [inline src/Rule.js]: right, so I'm asking can we get rid of `this.specificity` and `util/specificity` completely?

File: src/Style.js (hunk 19 above)
Comment (mikesherov) [inline src/Style.js]: but this constructor is always called with a value.

File: util/detectScoped.js [hunk 21]
```diff
@@ -1,59 +1,61 @@
...
+    // check if the selectorText has changed to the value we tried to set it to
+    // toLowerCase() it to account for browsers who change the text
+    var changeSelectorTextAllowed = 'd' === testStyle.selectorText.toLowerCase();
+
+    // remove the <style> to clean up
+    check.parentNode.removeChild(check);
+
+    // return the object with the appropriate flags
+    return {
+      scopeSupported: scopeSupported,
```
Comment (mikesherov) [inline util/detectScoped.js]: do we still need this?

File: util/detectScoped.js [hunk 22]
```diff
@@ -1,59 +1,61 @@
...
+      rules: DOMRules,
+      sheet: DOMStyle,
+      changeSelectorTextAllowed: changeSelectorTextAllowed
```
Comment (mikesherov) [inline util/detectScoped.js]: do we still need this?

File: util/detectScoped.js (hunk 6 above)
Comment (mikesherov) [inline util/detectScoped.js]: Where is this `check` being called now?

Comment (mikesherov) [inline util/detectScoped.js]: nvm, got it. 👍

File: util/detectScoped.js [hunk 23]
```diff
@@ -1,59 +1,61 @@
 // Scoped style support detection
```
Comment (mikesherov) [inline util/detectScoped.js]: comment now irrelevant, also please delete GH link

**Comment by mikesherov** [comment]
Good work, almost there! Really like the cleanliness of the new `Style` object.

File: src/Style.js (hunk 19 above)
Comment (kaicataldo) [inline src/Style.js]: Right, but it could be `null`. I put the default value in Style because that's the only thing that cares about `root` - would it be clearer if the default value happened at the Focss constructor level?

Comment (kaicataldo) [inline src/Style.js]: Ah, right, sorry, remembering why I put that there - it's because we can't reference `document` unless we're in the browser, and this encapsulates all of it within `Style`

Comment (mikesherov) [inline src/Style.js]: you mean `undefined`. You can pass undefined and it's initialize to `document.body`.... got it.

File: src/Rule.js (hunk 20 above)
Comment (kaicataldo) [inline src/Rule.js]: Oh I see - you're right, it's not used anywhere else

File: src/Style.js (hunk 19 above)
Comment (kaicataldo) [inline src/Style.js]: Yep!

**Comment by mikesherov** [comment]
👍

**Comment by mikesherov** [comment]
Breaking change, so this would be 2.0

**Comment by hzoo** [comment]
👍
